### ISSUE TITLE
(possible) _draw() function behaving weirdly when maximising and minimising window even WITHOUT randomise()

### ISSUE THREAD

**Issue by drmar** [godotengine/godot#37900]
**Godot version:**
3.2.1

**OS/device including version:**
macOS Catalin 10.15.4 / MacBookPro11,1

**Issue description:**
I have a Godot project with standard configuration. Nothing has been changed.
I wrote a simple line of code in func _draw():
draw_circle(Vector2(rand_range(0, 600), rand_range(0, 600)), rand_range(100, 300), Color(0.39, 0.58, 0.93, 1))
The (possible) problem: When I run the project, then maximise and minimise the window, the circle gets rearranged on the screen randomly even though I haven't used the randomise() function.
Check the screen recording for proof:
[Godot_Draw_Anomaly_without_randomise.mov.zip](https://github.com/godotengine/godot/files/4480434/Godot_Draw_Anomaly_without_randomise.mov.zip)

**Steps to reproduce:**
Use the attached Godot project.

**Minimal reproduction project:**
[Apocalypse Now_without_randomise.zip](https://github.com/godotengine/godot/files/4480423/Apocalypse.Now_without_randomise.zip)

**Comment by drmar** [comment]
Taking a closer look at the documentation, this is what it is saying about the draw() function:

> The _draw() function is only **called once**, and then the draw commands are cached and remembered, so further calls are unnecessary.

So, unless you are specifically telling Godot to redraw the shape by using the update() function, there is no reason for it to do so, especially when you are merely maximising and minimising the window.

**Comment by MWFIAE** [comment]
In my opinion a resize of the window would be a valid reason for the engine to redraw the window. 
So maybe that needs to be reflected in the dokumentation. 

Additional note: The rand_range does deliver random data even without calling randomise.
However the random numbers that will be returned would be the same every time a program is restarted. That's why randomise is desired most of the time.

**Comment by drmar** [comment]
@MWFIAE 

> In my opinion a resize of the window would be a valid reason for the engine to redraw the window.
So maybe that needs to be reflected in the dokumentation.

According to Godot's documentation, the draw() function is called only once, unless you specifically call the update() function, which I didn't.

 "The _draw() function is only called once, and then the draw commands are cached and remembered, so further calls are unnecessary.

If re-drawing is required because a state or something else changed, simply call CanvasItem.update() in that same node and a new _draw() call will happen."

Please check the documentation here: https://docs.godotengine.org/en/3.2/tutorials/2d/custom_drawing_in_2d.html

**Comment by MWFIAE** [comment]
@drmar 
Yeah I did understand you completely. 

That's why I sayed that maybe the documentation should be updated, if the statement "The _draw() function is only called once" isn't true (anymore). 
That's because I wouldn't consider the current behaviour a bug.

Also just for completion, I played around with your script and the following happens on a resize:
![image](https://user-images.githubusercontent.com/6072971/79381083-b72daa80-7f61-11ea-85da-fbd9111ee6e5.png)

The screen gets redrawn every time the size of the windows changes (even by a little bit). 
While the resize operation is ongoing ( I hold the button down) it isn't redrawn, except if I move the mouse.

**Comment by drmar** [comment]
@MWFIAE 

How about "draw commands are cached and remembered, so further calls are unnecessary."?
Wouldn't it make more sense and also be more efficient to draw everything from cache when the window is resized?
I mean, when you are resizing the window you are definitely not expecting, nor asking Godot to re-draw everything you have in the _draw() function. There may be cases where you have tens or hundreds of shapes. It doesn't make sense to re-draw everything every single time you resize the window.
In my opinion, the documentation is perfect. _draw() should be called only once, then cache everything. Only if you specifically ask Godot to re-draw everything in the _draw() function, by using update(), should it do it. Otherwise you end up with situations like this, when the engine is doing something it should and something you don't expect it to.

**Comment by Calinou** [comment]
@drmar Parts of the screen will have to be redrawn when resizing the window, as resizing may uncover parts that weren't drawn before.

Godot doesn't attempt to do partial drawing (it's too complex), so the method must be called again when the window is resized.

**Comment by drmar** [comment]
@Calinou 

> Parts of the screen will have to be redrawn when resizing the window, as resizing may uncover parts that weren't drawn before.

That's a good point. I haven't thought about that. I assumed everything was scaling with the window size.
This being the case, then I guess for the sake of efficiency only when making the window bigger should _draw() be called again? When you make the window smaller, there is no chance of leaving parts of the screen missing undrawn pixels. Just a thought.

Regarding documentation, so far I have only contributed to Godot by reporting issues, or at least situations that I believed to be an issue. I am not sure I am qualified to make edits to Godot's official documentation, but if you point me in the right direction I can learn how to do it. I am more than happy to help making Godot as god as possible.

**Comment by Calinou** [comment]
> When you make the window smaller, there is no chance of leaving parts of the screen missing undrawn pixels. Just a thought.

This is true only if you use the `disabled` stretch mode. If you use any other stretch mode, the drawing will have different coordinates and aspect ratio, which requires redrawing it.

**Comment by drmar** [comment]
@Calinou 

Alright. Thanks for the info! I'll certainly give it a try.

**Comment by AleryBerry** [comment]
I did not expect the _draw() function to be called when resizing windows, but it does. I think we should update the docs because it is not so clear.
